Feature Controls - Read Only Badge

kobelb opened the pull request:
# Read-Only Badges
To provide further reinforcement that a user is in a "read only" version of an application, we've added a badge to the upper right hand corner of the global nav bar:

<img width="249" alt="Screen Shot 2019-04-11 at 1 11 39 PM" src="https://user-images.githubusercontent.com/627123/55989976-6357f980-5c5b-11e9-95cf-49246e2f32c9.png">

# Platform Team

This required that we add a new core "badge" service. It's been mirrored after the breadcrumbs service. If you wouldn't mind giving this code a review, it'd be greatly appreciated

# Application Teams

For those of you using the angular's `uiRoutes`, a derivation of the following `badge` config was added:

```
uiRoutes
  .defaults(/dashboard/, {
    requireDefaultIndex: true,
    requireUICapability: 'dashboard.show',
    badge: (i18n, uiCapabilities) => {
      if (uiCapabilities.dashboard.showWriteControls) {
        return undefined;
      }

      return {
        text: i18n('kbn.dashboard.badge.readOnly.text', {
          defaultMessage: 'Read only',
        }),
        tooltip: i18n('kbn.dashboard.badge.readOnly.tooltip', {
          defaultMessage: 'Unable to save dashboards',
        }),
        iconType: 'glasses'
      };
    }
  })
```

For those of you not relying on angular's routing system, I tried to add the badge registration in the most consistent manner with how your alls applications handle this type of stuff. However, if you have a different preference, please let me know (PRs welcome as well).

kobelb committed:
Adding readonly badge to the discover application

kobelb committed:
Dashboard get a badge

kobelb committed:
Visualize gets a badge

kobelb committed:
Timelion gets a badge

kobelb committed:
Canvas gets a badge

kobelb committed:
Maps gets a badge

kobelb committed:
Infra gets a badge

kobelb committed:
Graph gets a badge

kobelb committed:
Dev Tools gets a badge

kobelb committed:
Index Patterns get badges

kobelb committed:
Advanced Settings get badges

kobelb committed:
Infra and i18n are super chill friends

kobelb committed:
Using proper i18n prefix for xpack

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8217/)

kobelb committed:
Adding badges to the uptime application

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8238/)

kobelb committed:
APM gets a badge!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8249/)

kobelb committed:
Adding functional tests for the discover read-only badge

kobelb committed:
Functional tests for everyone!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8271/)

kobelb committed:
Adding stubbed docs to get CI to run

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8274/)

kobelb committed:
Removing unused import

kobelb committed:
Fixing chrome service mock

kobelb committed:
Switching from ChromeBadge | null to ChromeBadge | undefined

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8286/)

kobelb committed:
Fixing canvas badge assertst

kobelb committed:
Fixing Logs ui capabilities

kobelb committed:
Merge remote-tracking branch 'upstream/granular-app-privileges' into fc/readonly-badge

kobelb committed:
Updating privileges api integration tests

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8296/)

kobelb committed:
More ChromeBrand | null to ChromeBrand | undefined related changes

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8305/)

kobelb committed:
Merge remote-tracking branch 'upstream/granular-app-privileges' into fc/readonly-badge

kobelb committed:
Fixing type error because of the merge

gchaps commented:
Can we add some context to the tooltip?

You're not authorized to save changes.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8321/)

kobelb commented:
> Can we add some context to the tooltip?

For sure, I'm working on the code changes to make this possible. 

> You're not authorized to save changes.

I like this verbiage much better, thanks!

kobelb committed:
Using named badges

kobelb commented on src/core/public/chrome/chrome_service.ts:
```diff
@@ -175,6 +183,29 @@ export class ChromeService {
+       */
+      getBadge$: (): Rx.Observable<ChromeBadge | undefined> =>
+        badge$.pipe(
+          map(badge => (typeof badge === 'string' ? namedBadges.get(badge) : badge)),
```
There's potential for us to make `namedBadges` to be an observable itself, so that if `addNamedBadge` is invoked after `setBadge` and `getBadge$` that it will at this time emit the proper badge. As it's written now, we'll just get `undefined` if this sequence of events occurs.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8347/)

legrego commented:
The one thing that bothers me is the fact that all the apps know enough to render a "readOnly" badge, but don't know how to do so. It's code in OSS that's only exercised by a hack in the security plugin. I know, you did exactly what I asked for...


**Caution:** brainstorming below

What would it look like if the UI Capability manipulators (currently: security & spaces) provided the "reason" for disabling a capability? I'm not sure what the data structure would look like, but this would allow an app to have all the information it needs to display the badge w/ tooltip, without security having to hack it in.

In our early talks about feature controls / UI Capabilities, we talked about having many reasons for disabling a capability -- not just security reasons. Having the reason/justification information alongside the `uiCapabilities` might help with that down the road (or: YAGNI...)

azasypkin commented:
> What would it look like if the UI Capability manipulators (currently: security & spaces) provided the "reason" for disabling a capability? I'm not sure what the data structure would look like, but this would allow an app to have all the information it needs to display the badge w/ tooltip, without security having to hack it in.

I like the idea (assuming it's not too expensive to implement). In addition to human-readable "reason" it'd be nice to have a "machine-readable" code (plugin-id that disabled capability + code within this plugin to avoid collisions,`{ source: 'security', code: .... }`?) as well so that apps can easily customize the badge if they need to, but the lack thereof isn't blocker at this point anyway.

kobelb commented:
@AlonaNadler @cchaos would you all mind weighing in on the current proposal? The plan we'd like to pursue is adding the "Read Only" badge to the global header without the tooltip before merging feature controls to `master` for 7.1. We'll need to make some moderately complex changes to the "ui capabilities" system to allow the security plugin to provide a reason why for it's disabling the "save" options to get the suggested tooltip.

kobelb committed:
Revert "Using named badges"

This reverts commit c0e341bee1cea5c5a42ebc4d4efedc8abed586b1.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8869/)

cchaos commented:
I'm not following as to why you can't have a generic tooltip for the read-only badge? And is that the only discussion point?

kobelb commented:
> I'm not following as to why you can't have a generic tooltip for the read-only badge? And is that the only discussion point?

We can have a generic tooltip for the badge. It'd have to be something like "You're unable to save dashboards", we can't mention "authorization" or "authority", etc.

cchaos commented:
Haha, yeah, well "You lack the authority" was not meant to be real.

kobelb commented:
@cchaos we'll also likely need some assistance/guidance from the design team on the best way to get the badges into the header and looking appropriate. The `.euiHeaderBookmarks` has `flex-grow: 1` which is causing the badge to be aligned to the right, and the badge isn't "vertically centered" in the header itself.

kobelb commented:
@gchaps in lieu of the above and us needing a more generic message for the read-only tooltip, how do you feel about: "Unable to save dashboards", etc. ? Is there a different phrase you'd recommend?

kobelb committed:
i18n'ing the uptime read-only badges

kobelb committed:
Adding ChromeService tests for badges

gchaps commented:
@kobelb I like "Unable to save dashboards."  Short and to the point.

cchaos commented on src/legacy/ui/public/chrome/directives/header_global_nav/components/header_badge.tsx:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    return (
+      <EuiBetaBadge
```
We've also decided to reduce it's footprint by using an icon instead of the full text like so:

<img width="240" alt="Screen Shot 2019-04-09 at 13 04 43 PM" src="https://user-images.githubusercontent.com/549577/55825614-53fa7400-5ad4-11e9-9ea2-fbe0004d4184.png">

But I have to get the glasses icon into EUI as it doesn't exist yet.

kobelb replied:
Awesome, thanks so much @cchaos!

cchaos replied:
https://github.com/elastic/kibana/pull/34824 is in Master with the glasses icon

kobelb replied:
That was 10x faster than I anticipated, thanks @cchaos!

kobelb committed:
Starting to add tests for the legacy badge API

kobelb committed:
Changing capitalization of "Read Only" to "Read only"

kobelb committed:
Adjusting styles

kobelb committed:
Adding $setupBadgeAutoClear tests

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8941/)

kobelb committed:
Changing the badge tooltip

kobelb committed:
Fixing timelion i18n prefix

kobelb committed:
Changing where Canvas sets the breadcrumbs

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9070/)

kobelb committed:
Merge remote-tracking branch 'upstream/granular-app-privileges' into fc/readonly-badge

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9222/)

kobelb committed:
Using a read-only badge with an icon

gchaps commented on src/legacy/core_plugins/timelion/public/app.js:
```diff
@@ -72,6 +72,20 @@ require('ui/routes')
+          defaultMessage: 'Read only',
+        }),
+        tooltip: i18n('timelion.badge.readOnly.tooltip', {
+          defaultMessage: 'Unable to save timelion sheets',
```
```suggestion
          defaultMessage: 'Unable to save Timelion sheets',
```

gchaps commented on x-pack/plugins/canvas/public/angular/controllers/canvas.js:
```diff
@@ -3,15 +3,30 @@
+            defaultMessage: 'Read only',
+          }),
+          tooltip: i18n.translate('xpack.canvas.badge.readOnly.tooltip', {
+            defaultMessage: 'Unable to save canvas workpads',
```
```suggestion
            defaultMessage: 'Unable to save Canvas workpads',
```

gchaps commented on src/legacy/core_plugins/kibana/public/discover/controllers/discover.js:
```diff
@@ -93,6 +93,20 @@ uiRoutes
+          defaultMessage: 'Read only',
+        }),
+        tooltip: i18n('kbn.discover.badge.readOnly.tooltip', {
+          defaultMessage: 'Unable to save saved searches',
```
Is the word "saved" necessary here?

kobelb replied:
We've referred to the actual "saved object type" elsewhere as "saved searches", but I do agree that this language is super awkward. I'm not opposed to "Unable to save searches"

gchaps committed:
Update x-pack/plugins/canvas/public/angular/controllers/canvas.js

Co-Authored-By: kobelb <brandon.kobel@gmail.com>

gchaps committed:
Update src/legacy/core_plugins/timelion/public/app.js

Co-Authored-By: kobelb <brandon.kobel@gmail.com>

kobelb committed:
Changing discover's read-only verbiage

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9246/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9261/)

kobelb committed:
Updating NP docs after adding an iconType to the ChromeBadge

kobelb committed:
Fixing issue I introduced with the rebasing

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9284/)

kobelb committed:
Removing tests for code that moved to an untested part of Kibana

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9290/)

cchaos commented:
@kobelb Can you fix the conflicts?

kobelb committed:
Merge remote-tracking branch 'upstream/granular-app-privileges' into fc/readonly-badge

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9304/)

cchaos commented on src/legacy/core_plugins/kibana/public/dashboard/index.js:
```diff
@@ -57,7 +57,22 @@ function createNewDashboardCtrl($scope, i18n) {
+        tooltip: i18n('kbn.dashboard.badge.readOnly.tooltip', {
+          defaultMessage: 'Unable to save dashboards',
+        }),
+        iconType: 'glasses'
```
Would it make it easier to maintain consistency if the iconType and text were global vars?

kobelb replied:
It definitely would. I played around with something similar [here](https://github.com/elastic/kibana/pull/34551/commits/c0e341bee1cea5c5a42ebc4d4efedc8abed586b1) but it was later reverted because it added some indirection.

cchaos commented on src/legacy/ui/public/chrome/api/badge.test.ts:
```diff
@@ -0,0 +1,64 @@
+      const { chrome } = setup();
+      const badge = {
+        text: 'foo',
+        tooltip: `foo's tooltip`,
```
Should this also be checking for the iconType?

cchaos commented on src/legacy/ui/public/chrome/directives/header_global_nav/components/header_badge.tsx:
```diff
@@ -0,0 +1,94 @@
+      <div
+        className="chrHeaderBadge__wrapper"
+        data-test-subj="headerBadge"
+        aria-label={this.state.badge.text}
```
This aria-label is no longer needed since the BetaBadge connects the aria attributes with the tooltip.

<img width="609" alt="Screen Shot 2019-04-12 at 10 39 15 AM" src="https://user-images.githubusercontent.com/549577/56045497-598fce00-5d0f-11e9-9a23-742be51aae16.png">

kobelb replied:
I'm only seeing the euiToolTipPopover in the DOM when hovering over the tooltip. I primarily added this for use within the [global nav functional testing service](https://github.com/elastic/kibana/pull/34551/files#diff-5841023215c87b8fa2d1ff1fd103ba65R49). If we're confident that screenreaders behave properly without this `aria-label` itself, I can switch to something like `data-test-badgeText`.

cchaos replied:
It also shows in the DOM on focus so when users tab to the item, it will appear:

<img src="https://d.pr/free/i/xhvSAR+" width="50%" />

cchaos commented on x-pack/plugins/graph/public/badge.js:
```diff
@@ -0,0 +1,22 @@
+      defaultMessage: 'Read only',
+    }),
+    tooltip: i18n('xpack.graph.badge.readOnly.tooltip', {
+      defaultMessage: 'Unable to save graph workspaces',
```
```suggestion
      defaultMessage: 'Unable to save Graph workspaces',
```

cchaos reviewed (approved):
This works as described and prescribed (checked in FF and IE too). Just had a couple of suggestions.

bmcconaghy reviewed (approved):
console part LGTM

w33ble reviewed (approved):
Canvas looks good 🚀

lukeelmers reviewed (approved):
Tested dashboard (+ dashboard only mode), discover, index patterns, settings, visualize, timelion... all LGTM!

Bargs reviewed (approved):
Since Luke said he tested Discover I trust him 😉

kobelb commented:
This PR was against the `granular-app-privileges` branch, which already merged to master last week. I've rebased the branch against master and fixed some issues, it's available here: https://github.com/elastic/kibana/pull/35252

For those of you who have already reviewed this PR, and there weren't any new changes, I haven't requested your review on the new rebased PR.
